### ISSUE TITLE
Performance issues with String#match? and Regexp#match?

### ISSUE THREAD

**Issue by ashmaroli** [marcandre/backports#117]
I benchmarked `String#match?` and `Regexp#match?` backports against `String#=~` on Ruby 2.3 and Ruby 2.4 to assess their performance scores..

### Script
```ruby
require 'benchmark/ips'
require 'backports/2.4.0/string/match'
require 'backports/2.4.0/regexp/match'

SLUG = 'test_some_kind_of_long_file_name.rb'

def slower
  SLUG =~ /^test_/
end

def slow
  SLUG.match?(/^test_/)
end

def equally_slow
  /^test_/.match?(SLUG)
end

Benchmark.ips do |x|
  x.report('String#=~')     { slower }
  x.report('String#match?') { slow }
  x.report('Regexp#match?') { equally_slow }
  x.compare!
end
```

### Ruby 2.3.1 x86 Windows
```
Warming up --------------------------------------
           String#=~    31.000k i/100ms
       String#match?     2.822k i/100ms
       Regexp#match?     2.937k i/100ms
Calculating -------------------------------------
           String#=~    563.641k (A± 3.4%) i/s -      2.821M in   5.011029s
       String#match?     30.934k (A±15.5%) i/s -    152.388k in   5.065203s
       Regexp#match?     30.119k (A±16.0%) i/s -    146.850k in   5.021743s

Comparison:
           String#=~:   563641.4 i/s
       String#match?:    30933.5 i/s - 18.22x  slower
       Regexp#match?:    30118.8 i/s - 18.71x  slower
```

### Ruby 2.4.4 x64 Windows
```
Warming up --------------------------------------
           String#=~    41.998k i/100ms
       String#match?    72.822k i/100ms
       Regexp#match?    71.965k i/100ms
Calculating -------------------------------------
           String#=~    667.187k (A± 2.2%) i/s -      3.360M in   5.038405s
       String#match?      1.914M (A± 4.6%) i/s -      9.540M in   5.001935s
       Regexp#match?      1.943M (A± 1.1%) i/s -      9.715M in   5.000873s

Comparison:
       Regexp#match?:  1942953.3 i/s
       String#match?:  1913649.2 i/s - same-ish: difference falls within error
           String#=~:   667186.9 i/s - 2.91x  slower
```

I was ready to see a performance issue between the two because:
  - Ruby 2.4 is a 64-bit installation vs 32-bit Ruby 2.3
  - `:match?` in Ruby 2.4 is natively defined vs Ruby `;match?` backport
but did not expect to see such a large difference..

Could the performance differences be due to above factors or because the backport *initializes a `Fiber`*, or both?

**Comment by marcandre** [comment]
Indeed, `match?` is difficult to backport and we had to resort to using a Fiber to not touch the magic variables `$1`, etc.

What we didn't check was if doing `Fiber.new` only once and storing the result would help performance. My guess is that it won't, but I might be wrong. PR welcome in case it does :-)

**Comment by marcandre** [comment]
I also added a note to the README about that fact that `match?` is quite a bit slower than all other backports.

**Comment by Arcovion** [comment]
Couldn't you just restore `$~`?
```ruby
last_matchdata = $~
!match(*args).nil?
$~ = last_matchdata
```
`$&` is an alias for `$~.string` and `$1-9` are aliases for `$~[1-9]` so they work fine once `$~` is changed back.

**Comment by marcandre** [comment]
Wow, I never realized that `$~` was writable.
Sure, that should work fine and be much faster, thanks!

**Comment by ashmaroli** [comment]
Yep, @Arcovion's solution is faster:
```
Warming up --------------------------------------
           String#=~    30.714k i/100ms
       String#match?    21.786k i/100ms
       Regexp#match?    23.741k i/100ms
Calculating -------------------------------------
           String#=~    569.452k (A± 2.3%) i/s -      2.856M in   5.018670s
       String#match?    343.463k (A± 1.1%) i/s -      1.721M in   5.011547s
       Regexp#match?    397.319k (A± 1.2%) i/s -      1.994M in   5.019943s

Comparison:
           String#=~:   569452.3 i/s
       Regexp#match?:   397318.7 i/s - 1.43x  slower
       String#match?:   343462.6 i/s - 1.66x  slower
```

**Comment by marcandre** [comment]
I have a fix almost ready, but there's some strange issue with the specs (they don't fail on a naive implementatino) and I have to go, so I'll fix this later today.

**Comment by ashmaroli** [comment]
> so I'll fix this later today.

sure.. no hurry..

**Comment by marcandre** [comment]
Turns out, `Regexp.last_match` is method local, so there's not even a need to restore it!
Released 3.11.4
Thanks to all :-)
